### PR TITLE
char (*)[] -> char ** for clang compatibility

### PR DESCRIPTION
**Pull request by MaskRay**
Otherwise,
error: dereference of pointer to incomplete type 'const char []'

### CODE REVIEW

**Comment by lhmouse** [comment]
The `"o"` constraints are indispensable for informing GCC that memory locations near their operands (which must be _lvalues_ despite incompleteness) are to be read from or written to. The necessity can be observed in [this example](https://godbolt.org/g/MCXgK4).

Note, again, that `"o"` accept an lvalue of _the to-be-accessed operand_, so having it take a `char *` is definitely  incorrect, since we aren't modifying a `char *`.

If dereferencing a pointer to an array of unknown bound will not be accepted by clang, you have to work it around by declaring the `__asm__` statement `volatile` then adding a `"memory"` clobber. This prevents reordering of all memory access, including those irrelevant to the locations being accessed. It is consequently discouraged and superseded.

File: src/env/inline_mem.h [hunk 1]
```diff
@@ -18,15 +18,15 @@ __attribute__((__always_inline__)) static inline void * _MCFCRT_inline_mempcpy_f
...
-			: "=o"(*(char (*)[])__p1), "+D"(__p1), "+S"(__p2), "=c"(__dumb)
-			: "o"(*(const char (*)[])__p2), "c"(__n / 8)
+			: "=o"(*(char **)__p1), "+D"(__p1), "+S"(__p2), "=c"(__dumb)
+			: "o"(*(const char **)__p2), "c"(__n / 8)
 		);
 	}
 	if(__builtin_constant_p(__n) ? (__n % 8) : true){
 		__asm__ (
 			"rep movsb \n"
 			: "=o"(*(char (*)[])__p1), "+D"(__p1), "+S"(__p2), "=c"(__dumb)
```
Comment (FrankHB) [inline src/env/inline_mem.h]: Seems missing modification of this line?

Comment (FrankHB) [reply src/env/inline_mem.h]: To be clear, I meant the line 28.

Comment (lhmouse) [reply src/env/inline_mem.h]: The extended asm constraints has no syntax for ranged parameters, so I am afraid there is no solution other than a `"memory"` clobber.

**Comment by lhmouse** [comment]
I would have closed this as a duplicate of #18. :joy:

**Comment by lhmouse** [comment]
There are other considerations: Clang [doesn't implement ASM constraints correctly](https://godbolt.org/g/yPMUxB). It does not read Intel syntax, either.

**Comment by lhmouse** [comment]
I am closing this as a duplicate of #18.

**Comment by lhmouse** [comment]
Before the formal initiation of clang support development none of these changes will be adopted.

**Comment by FrankHB** [comment]
Any suggestion from Clang folks?

**Pull request closed without merging**